3.13/redundancy_eliminator.md

```diff
@@ -0,0 +1,86 @@
+# The Tier 2 Redundancy Eliminator
+
+Author: Ken Jin, Mark Shannon
+
+Tier 2 traces require optimization. At minimum, redundant operations
+should be removed. This is where the name "Redundancy Eliminator" is
+derived from. This forms part of the larger effort towards
+the tier 2 optimizer.
+
+## Abstract Interpretation of Uops
+
+To analyze the uops, [abstract interpretation](https://en.wikipedia.org/wiki/Abstract_interpretation)
+shall be used.
+
+Just like the CPython interpreter is an interpretation over
+Python program state, represented by `PyObject *`, the abstract interpreter
+shall be an interpretation over Python symbolic information state,
+represented by `_Py_UOpsSymType *`.
+
+### Lattice
+
+The states of information in such an abstract interpretation can be
+represented by a [lattice](https://en.wikipedia.org/wiki/Lattice_(order))
+of values. The lattice itself is an ordered powerset
+of all variables and their associated information.
+
+For example, here is a trivial lattice for the statement 
+```python
+x = 1
+y = 2.0
+```
+:
+
+```mermaid
+flowchart BT
+    A("`Bottom, ie. x: *?*, y: *?*`") --> B("`x: *int*, y: *?*`")
+    A --> C("`x: *?*, y: *float*`")
+    B --> D("`Top, x: *int*, y: *float*`")
+    C --> D
+  
+```
+
+This example is intentionally contrived. Real Python runtime information
+may contain more complex types, constants, and unions of types.
+
+Each `_Py_UOpsSymType` thus captures all the information associated
+with a variable at any point in the lattice.
+
+### Abstract DSL Generator
```

Comment:
Remember that "DSL" stands for "domain-specific language". We refer to the grammar of bytecodes.c as a DSL because it has a custom grammar (not standard C or Python) for which we wrote a custom parser.

But "DSL generator" sounds like it generates the DSL. Our generator doesn't -- it rather takes a file written in our Domain Specific Language and generates (most of) an abstract interpreter.